### PR TITLE
concurrent_percolator: remove the cost of getting commitTS

### PR DESCRIPTION
**Pull request by alivxxx**
The commit ts now is `max(start_ts, key_last_read_ts[k1], key_last_read_ts[k2]....) + 1`.
PTAL @coocood @siddontang @GregoryIan

### CODE REVIEW

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 1]
```diff
@@ -185,9 +185,12 @@ canLockKey(k, ts) ==
     /\ writes = {}       \* no any newer writes or rollbacks.
 
 \* Locks the key and places the data.
-lockKey(k, start_ts, primary) ==
+lockKey(c, k, start_ts, primary) ==
   /\ key_lock' = [key_lock EXCEPT ![k] = @ \union {[ts |-> start_ts, primary |-> primary]}]
   /\ key_data' = [key_data EXCEPT ![k] = @ \union {[ts |-> start_ts]}]
```
Comment (foreverbell) [inline ConcurrentPercolator/ConcurrentPercolator.tla:190]: L48 states that `key_last_read_ts` is actually an auxiliary variable, but looks like now it should exist in real world implementation, can we fix that comment?

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 2]
```diff
@@ -185,9 +185,12 @@ canLockKey(k, ts) ==
...
+  /\ IF client_ts[c].commit_ts < key_last_read_ts[k]
```
Comment (foreverbell) [inline ConcurrentPercolator/ConcurrentPercolator.tla:191]: I prefer adding a utility function named `Max`.

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 3]
```diff
@@ -251,10 +254,9 @@ Prewrite(c) ==
   /\ client_state[c] = "prewriting"
   /\ IF client_key[c].pending = {}
      THEN  \* all keys have been pre-written
-       /\ next_ts' = next_ts + 1
        /\ client_state' = [client_state EXCEPT ![c] = "committing"]
-       /\ client_ts' = [client_ts EXCEPT ![c].commit_ts = next_ts']
-       /\ UNCHANGED <<key_vars, client_key>>
+       /\ client_ts' = [client_ts EXCEPT ![c].commit_ts = client_ts[c].commit_ts + 1]
```
Comment (foreverbell) [inline ConcurrentPercolator/ConcurrentPercolator.tla:258]: Comment something like, `client_ts[c].commit_ts` is set to `start_ts` at first, then updated at every prewrite, and finally increased by 1.

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 4]
```diff
@@ -14,7 +14,7 @@ CONSTANTS CLIENT_PRIMARY_KEY
 ASSUME CLIENT_PRIMARY_KEY \in [CLIENT -> KEY]
 
 \* $next_ts$ is the timestamp for transaction. It is increased monotonically,
-\* so every transaction must have a unique start and commit ts.
+\* so every transaction must have a unique start ts.
```
Comment (foreverbell) [inline ConcurrentPercolator/ConcurrentPercolator.tla:17]: .. so it is allowed that `txnA.start_ts = txnB.commit_ts` ?

File: ConcurrentPercolator/ConcurrentPercolator.tla [hunk 5]
```diff
@@ -414,7 +416,7 @@ AbortedConsistency ==
     (/\ client_state[c] = "aborted"
      /\ client_ts[c].commit_ts # 0
```
Comment (foreverbell) [inline ConcurrentPercolator/ConcurrentPercolator.tla:417]: Looks like this premise (`client_ts[c].commit_ts # 0`) is no longer an valid one.

**Review by foreverbell — COMMENTED** [review]
Since I write this code, so let me review it first :)

I think the concurrent feature is orthogonal to this feature (I would call it smart commit_ts or something else), so it would be better for us to branch one based on the version in `Percolator` folder, rather than modifying this version directly.

**Comment by siddontang** [comment]
I prefer using another TLA+ file instead of modifying directly. So we can trace every modification is correct.

File: ConcurrentPercolator/ConcurrentPercolator.tla (hunk 1 above)
Comment (alivxxx) [reply ConcurrentPercolator/ConcurrentPercolator.tla:190]: No, in the real world, the `last_read_ts` only exists in the region level, not key level.

File: ConcurrentPercolator/ConcurrentPercolator.tla (hunk 5 above)
Comment (alivxxx) [reply ConcurrentPercolator/ConcurrentPercolator.tla:417]: Why? If the previous state is `init`, the `commit_ts` is 0.

File: ConcurrentPercolator/ConcurrentPercolator.tla (hunk 4 above)
Comment (alivxxx) [reply ConcurrentPercolator/ConcurrentPercolator.tla:17]: Yes, it could happen.

File: ConcurrentPercolator/ConcurrentPercolator.tla (hunk 1 above)
Comment (foreverbell) [reply ConcurrentPercolator/ConcurrentPercolator.tla:190]: By saying something is auxiliary, I mean it corresponds to nothing in a real-world implementation. We add it to make some checking doable. Hereby, `last_read_ts` does have a mapping to **something** in implementation, though it is a little different. But remember TLA+ spec is something abstract, we don't have a region concept here, so saying  `last_read_ts` exists in real-world implementation is okay, IMHO.

**Pull request closed without merging**